File: ios/FLEXR/Data/Repository/MatchRepository.swift
import Foundation

/// Match-Liste mit lokalem Bestand als Single Source of Truth.
///
/// Die Oberfläche liest ausschließlich aus [FlexrStore], das Netz füllt nach.
/// Dadurch sind Matches und Chats sofort und auch offline sichtbar — im Web war
/// jede Ansicht ein Ladebalken.
@MainActor
@Observable
final class MatchRepository {

    /// Aktueller Stand aus dem lokalen Bestand, nie direkt vom Server.
    private(set) var matches: [MatchSummary] = []

    /// Der Menüpunkt „Chats".
    ///
    /// Bewusst `inChats` statt `lastMessage != nil`: Nach „Chatverlauf leeren"
    /// liefert der Server korrekt `last_message = null`, der Chat soll aber
    /// gelistet bleiben — nur eben leer. Erst „Chat löschen" setzt `in_chats`
    /// zurück, bis eine neue Nachricht eintrifft.
    var conversations: [MatchSummary] { matches.filter(\.inChats) }

    var unreadTotal: Int { matches.reduce(0) { $0 + $1.unreadCount } }

    @ObservationIgnored private let api: FlexrAPI
    @ObservationIgnored private let store: FlexrStore

    init(api: FlexrAPI, store: FlexrStore) {
        self.api = api
        self.store = store
        reload()
    }

    func match(id: String) -> MatchSummary? {
        matches.first { $0.matchID == id }
    }

    @discardableResult
    func refresh() async throws -> [MatchSummary] {
        let remote = try await api.matches().map { $0.toDomain() }
        store.replaceMatches(remote)
        reload()
        return remote
    }

    /// Ungelesen-Zähler lokal zurücksetzen, sobald ein Chat geöffnet wurde.
    func markRead(matchID: String) {
        store.clearUnread(matchID: matchID)
        reload()
    }

    /// Match auflösen: Match und Chatverlauf werden serverseitig gelöscht, der
    /// eigene Swipe ebenfalls — die Person erscheint dadurch erneut im Deck.
    /// Eine Sperre wie beim Blockieren ist das ausdrücklich nicht.
    func unmatch(matchID: String) async throws {
        try await api.unmatch(matchID: matchID)
        store.deleteMatch(id: matchID)
        reload()
    }

    /// „Chat löschen": anders als [unmatch] bleibt das Match serverseitig
    /// bestehen — nur die Unterhaltung verschwindet aus dem „Chats"-Tab, bis
    /// erneut eine Nachricht eintrifft. Lokal reicht ein [refresh], das den
    /// jetzt aktualisierten `inChats`-Wert vom Server übernimmt.
    func deleteChat(matchID: String) async throws {
        try await api.deleteChat(matchID: matchID)
        store.deleteMessages(matchID: matchID)
        try await refresh()
    }

    /// Nach dem Blockieren verschwindet das Match beidseitig aus der Liste.
    func removeLocally(matchID: String) {
        store.deleteMatch(id: matchID)
        reload()
    }

    func clearLocalCache() {
        store.deleteAll()
        reload()
    }

    private func reload() {
        matches = store.allMatches().map { $0.toDomain() }
    }
}

File: android-native/app/src/main/java/flexr/social/app/data/repository/MatchRepository.kt
package flexr.social.app.data.repository

import flexr.social.app.core.network.apiCall
import flexr.social.app.data.local.MatchDao
import flexr.social.app.data.local.MessageDao
import flexr.social.app.data.local.toDomain
import flexr.social.app.data.local.toEntity
import flexr.social.app.data.remote.FlexrApi
import flexr.social.app.domain.model.MatchSummary
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Match-Liste mit lokalem Cache als Single Source of Truth: die Oberfläche
 * liest ausschließlich aus Room, das Netz füllt den Cache nach. Dadurch sind
 * Matches und Chats sofort und auch offline sichtbar — im Web war jede
 * Ansicht ein Ladebalken.
 */
@Singleton
class MatchRepository @Inject constructor(
    private val api: FlexrApi,
    private val matchDao: MatchDao,
    private val messageDao: MessageDao,
) {

    val matches: Flow<List<MatchSummary>> =
        matchDao.observeAll().map { rows -> rows.map { it.toDomain() } }

    /** Nur Matches mit laufender Unterhaltung — der Menüpunkt „Chats". */
    val conversations: Flow<List<MatchSummary>> =
        matchDao.observeWithConversation().map { rows -> rows.map { it.toDomain() } }

    val unreadTotal: Flow<Int> = matchDao.observeUnreadTotal()

    fun match(matchId: String): Flow<MatchSummary?> =
        matchDao.observeById(matchId).map { it?.toDomain() }

    suspend fun findMatch(matchId: String): MatchSummary? = matchDao.findById(matchId)?.toDomain()

    suspend fun refresh(): List<MatchSummary> {
        val remote = apiCall { api.getMatches() }.map { it.toDomain() }
        matchDao.replaceAll(remote.map { it.toEntity() })
        return remote
    }

    /** Ungelesen-Zähler lokal zurücksetzen, sobald ein Chat geöffnet wurde. */
    suspend fun markRead(matchId: String) = matchDao.clearUnread(matchId)

    /**
     * Match auflösen: Match und Chatverlauf werden serverseitig gelöscht, der
     * eigene Swipe ebenfalls — die Person erscheint dadurch erneut im Deck.
     * Eine Sperre wie beim Blockieren ist das ausdrücklich nicht.
     */
    suspend fun unmatch(matchId: String) {
        apiCall { api.unmatch(matchId) }
        messageDao.deleteForMatch(matchId)
        matchDao.deleteById(matchId)
    }

    /**
     * "Chat löschen": anders als [unmatch] bleibt das Match serverseitig
     * bestehen - nur die Unterhaltung verschwindet aus dem "Chats"-Tab, bis
     * erneut eine Nachricht eintrifft. Lokal reicht ein [refresh], das den
     * jetzt aktualisierten inChats-Wert vom Server übernimmt.
     */
    suspend fun deleteChat(matchId: String) {
        apiCall { api.deleteChat(matchId) }
        messageDao.deleteForMatch(matchId)
        refresh()
    }

    /** Nach dem Blockieren verschwindet das Match beidseitig aus der Liste. */
    suspend fun removeLocally(matchId: String) {
        messageDao.deleteForMatch(matchId)
        matchDao.deleteById(matchId)
    }
}
